### ISSUE TITLE
Outputting the parsed AST

### ISSUE THREAD

**Issue by sjakobi** [haskell-nix/hnix#661]
I'd like to better understand how `hnix` represents the parsed source – ideally I'd like to see the `Show` output for the parsed `NExpr`.

What's currently the best way to do that? Run `parseNixText` in `ghci`?

`dhall` has a useful `haskell-syntax-tree` command for this. Example:

```
$ echo 'if Natural/odd x then "odd" else Natural/show x' | dhall haskell-syntax-tree
BoolIf
    ( App NaturalOdd
        ( Var ( V "x" 0 ) )
    )
    ( TextLit ( Chunks [] "odd" ) )
    ( App NaturalShow
        ( Var ( V "x" 0 ) )
    )
```

Would that be something worth replicating in `hnix`?

**Comment by sorki** [comment]
Yup, or `parseNixTextLoc` -> to debug these I wrote a little helper recently:

```haskell
dbgExpr :: NExprLoc -> a
dbgExpr e = error
  $ Data.Text.unpack
  $ Data.Text.Lazy.toStrict
  $ Text.Pretty.Simple.pShow
  $ Nix.stripAnnotation e
```

Would be nice to have this as a part of CLI (or REPL) for sure.

(Btw I need a little help with GitHub magic - how can you embed a snippet of code in GitHub from another repo?  :mage: )

**Comment by sjakobi** [comment]
Oh, that looks like a nice helper!

> (Btw I need a little help with GitHub magic - how can you embed a snippet of code in GitHub from another repo?  )

No idea! If it's supported, I'd like to know too! :)

**Comment by sorki** [comment]
> Oh, that looks like a nice helper!
> 
> > (Btw I need a little help with GitHub magic - how can you embed a snippet of code in GitHub from another repo?  )
> 
> No idea! If it's supported, I'd like to know too! :)

Was looking at these https://github.com/haskell-nix/hnix/issues/659#issuecomment-650112401 the other day and wondering :) Maybe they just need to be on a separate line or maybe it just works for current repo only and not cross-repo.

**Comment by sjakobi** [comment]
So far it has worked for me only within the same repo.

**Comment by Anton-Latukha** [comment]
Maybe currently also these ideas would help.

This is probably a more fitting example of `pretty-simple` use. I thought maybe we can offload parts of REPL output onto it, but seems much more direct case.

I also want to learn to use http://felsin9.de/nnis/ghc-vis/, it would not print the Nix AST, but it is already quite something.

**Comment by sjakobi** [comment]
> This is probably a more fitting example of `pretty-simple` use.

Absolutely! In fact, both `dhall haskell-syntax-tree` and @sorki's `dbgExpr` use `pretty-simple`! :)

**Comment by Anton-Latukha** [comment]
> (Btw I need a little help with GitHub magic - how can you embed a snippet of code in GitHub from another repo? mage )

Test:
https://github.com/Anton-Latukha/haskell-with-nixpkgs/blob/d1fa183f4769787ba116b8a119a9bed009a5c45f/default.nix#L1-L100

**Comment by Anton-Latukha** [comment]
https://github.com/NixOS/nix/blob/0b606aad46e1d96da36d4831df63ad90f11d21c3/src/libstore/build.cc#L491-L499

**Comment by Anton-Latukha** [comment]
Yes, indeed, the Show for AST is essential feature for HNix project development.

It should be somehow thought since HNix AST most probably sometimes would be infinite. So at least it should Weak Head Normal Form Show the levels of recursion until user stops the output, or something like that, otherwise AST can become just `foldl` kind of traverse that never traverses the AST and never terminates and so does not print anything.

I never heard the explicit claim that "Nix language is designed so it always terminates", so that why I consider that nevertheless there are recursion checks - infinite recursions is most probably possible. Especially in our current data type system, moreover having it in recursive schemes.terminate 

Or well, for usefulness and brevity - at least we can merge the https://github.com/haskell-nix/hnix/issues/661#issuecomment-650584488 sort of solution and terminate the HNix by hand or by OOM Killer if it goes into infinity. In reality - that is effective and no biggy, especially at this stage - the main thing is to be able to print the AST for us to understand it better.

**Comment by Anton-Latukha** [comment]
For example:

Case:
* AST has < 11 levels - parse and print the whole AST in the consistent and pretty form by printing it all through `pretty-simple`, that job should terminate still pretty quickly.

* AST > 10 - go into showing AST by showing layer (L1) and abstractions it uses, then below start showing L1.abstractions (L2), and then L1.L2.abstractions (L3), layer by layer by layer... That would still be readable and trackable to follow.

**Comment by sjakobi** [comment]
> It should be somehow thought since HNix AST most probably sometimes would be infinite. So at least it should Weak Head Normal Form Show the levels of recursion until user stops the output, or something like that, otherwise AST can become just `foldl` kind of traverse that never traverses the AST and never terminates and so does not print anything.
> 
> I never heard the explicit claim that "Nix language is designed so it always terminates", so that why I consider that nevertheless there are recursion checks - infinite recursions is most probably possible. Especially in our current data type system, moreover having it in recursive schemes.terminate

I don't understand the implementation very well at this stage, but I wouldn't expect the parsed AST to contain loops. I suspect that any cycles would be introduced later during evaluation.

**Comment by Anton-Latukha** [comment]
Currently `[nix| expr |]` QuasiQuoter shows the AST.
Example: https://github.com/haskell-nix/hnix/issues/52#issuecomment-955608351

But you are right - the `show` functions are not that obvious, and some of them are hidden as local functions in the source code. And which are exposed - are not documented as such, and are in the `Pretty` module
```haskell
-- ./Nix/Pretty.hs
prettyNix :: NExpr -> Doc ann
printNix :: forall t f m . MonadDataContext f m => NValue t f m -> Text
````

As I think, they not only should be documented, but also renamed & probably relocated, or exposed in main modules.

**Comment by Anton-Latukha** [comment]
https://github.com/haskell-nix/hnix/issues/661#issuecomment-650584488
> (Btw I need a little help with GitHub magic - how can you embed a snippet of code in GitHub from another repo? mage )

Answer is: there is none. Any kinds of links to code in other repos are always represented as a link.
